genetic-analysis as link between Event, Material, Protocol, Sequence

thomasstjerne opened the pull request:
This should reflect the model discussed at the dwc-dp meeting.
I have not changed the naming "genetic-....." , but I think we should consider that.

This introduces the join/link table `GeneticAnalysis` that links events, protocols, sequences and optionally material.

In order to re-use `GeneticAnalysis` for targeted assays, geneticSequenceID is nullable. In targeted assays, Identifications should use `identificationBasedOnGeneticAnalysisID` to point directly to the analysis, whereas Identifications for metabarcoding should use `identificationBasedOnGeneticSequenceID`.

The `GeneticProtocol` is basically the current DNA_derived_data extension, with the exclusion of the following fields that are better placed in other entities:

````
occurrenceID
samp_name
project_name
samp_taxon_id
env_broad_scale
env_local_scale
env_medium
subspecf_gen_lin
ploidy
biotic_relationship
specific_host
host_spec_range
host_disease_stat
trophic_level
rel_to_oxygen
tax_class
otu_class_appr
otu_seq_comp_appr
otu_db
````

We might also consider whether the following fields would be better placed elsewhere:

`````
propagation
samp_collec_device
samp_collec_method
tax_ident
ref_db
num_replicons
extrachrom_elements
pathogenicity
compl_score
````

timrobertson100 commented on sandbox/experimental/data-packages/dwc-dp/0.1/table-schemas/genetic-analysis.json:
```diff
@@ -0,0 +1,128 @@
+  "description": "A link between a DNA sequence and the event and material from which it was derived, using a specified protocol",
+  "fields": [
+    {
+      "name": "geneticAnalysisID",
```
Do you normally have such an ID? If not, a compound PK across eventID, geneticProtocolID, geneticSequenceID, materialID could be used.

thomasstjerne replied:
We don´t. But we were trying to also accommodate targeted assays (pcr) where the sequence does not exist. Then we need to be able to link directly from `Identification` through `identificationBasedOnGeneticAnalysisID`.

timrobertson100 replied:
Ah of course - thanks. Wasn't thinking this through far enough.

tucotuco replied:
And without the unique ID, it there wouldn't be a consistent way to link Assertions to it.

timrobertson100 commented on sandbox/experimental/data-packages/dwc-dp/0.1/table-schemas/genetic-sequence.json:
```diff
@@ -84,14 +40,5 @@
-            }
-        }
-    ]
+    "primaryKey": "geneticSequenceID"
```
Is it worth having a unique on the sequence as well as the ID?

tucotuco replied:
With the new concept of what this table is for, the geneticSequenceID as an arbitary ID actually makes no sense, as it would not function in aggregation. The global unique identifier should be the MD5 of the sequence itself. The term geneticSequenceID should remain, but it should be a requirement that it be the MD5 of the sequence. Then the question is, how would a geneticSequenceRemarks ever be used? That wouldn't work well in aggregation either. That leaves GeneticSequence as a two field schema.

CecSve replied:
The MD5 is (I believe) specifically for the sequencing technology, and I think you may have cases where it is `OTU1` or `OTU_1`. Nevertheless, it should always be unique in this table. We should not include the MD5 requirement in the description.

timrobertson100 commented on sandbox/experimental/data-packages/dwc-dp/0.1/index.json:
```diff
@@ -186,13 +200,6 @@
             "title": "GeneticSequenceMedia",
```
I suggest to remove Media here and in the description. Simply state that this is the GeneticSequence.

tucotuco replied:
I disagree. This is a join table between GeneticSequence and Media, so that GeneticSequences can appear in Media. It follows the pattern of all [Class]Media join tables.

timrobertson100 replied:
You are correct. It's the github view that has hidden lines, making it appear under the http://rs.gbif.org/sandbox/experimental/data-packages/dwc-dp/0.1/table-schemas/genetic-sequence.json URL

tucotuco reviewed (commented):
I don't see here the recommendations for Identification. Talking with @CecSve about that, the otu_db map to identificationReferences, with appropriate adjustments to the definition, usage comments, and examples. The otu_seq_comp_appr could be incorporated as something like taxonAssignmentMethod.

Other than these observations, I think it should be fine to merge this.

thomasstjerne commented:
I have added to `Identification`:

1.  `identificationBasedOnGeneticAnalysisID`. A foreign key directly to  `GeneticAnalysis`. This will allow for qPCR, ddPCR and shotgun sequencing where there would not be `GeneticSequence` objects in the data.
2. `taxonAssignmentMethod`. The example and description  I have added are very specific for DNA data, based on MiXS terms `otu_class_appr` and `otu_seq_comp_appr`. I will leave to others to expand this to other use cases for this field.

mdoering commented:
> * `identificationBasedOnGeneticAnalysisID`. A foreign key directly to  `GeneticAnalysis`. This will allow for qPCR, ddPCR and shotgun sequencing where there would not be `GeneticSequence` objects in the data.

does it need to be so verbose or could `geneticAnalysisID` be enough? And similar just `geneticSequenceID`?

tucotuco commented:
I don't believe I am allowed to approve nor merge, but I am fine with that being done if there are no more pending changes.

timrobertson100 commented:
Based on that, I'll merge it. We can adjust if needed.

timrobertson100 commented:
```
Error: Unreachable URL: http://rs.gbif.org/sandbox/experimental/data-packages/dwc-dp/0.1/table-schemas/genetic-sequence-protocol.json
Error: File does not exist sandbox/experimental/data-packages/dwc-dp/0.1/table-schemas/genetic-sequence-protocol.json
```

tucotuco commented:
> ```
> Error: Unreachable URL: http://rs.gbif.org/sandbox/experimental/data-packages/dwc-dp/0.1/table-schemas/genetic-sequence-protocol.json
> Error: File does not exist sandbox/experimental/data-packages/dwc-dp/0.1/table-schemas/genetic-sequence-protocol.json
> ```

I'll take care of this since it has been merged...

timrobertson100 commented:
https://github.com/gbif/rs.gbif.org/pull/180/commits/97e75b2942e659a24311c5a9e9191814c9ef69f9
